internal/types/document_validation.go

```diff
@@ -85,31 +89,66 @@ func (d *Document) ValidateData() error {
 		}
 		duplicateChecker[key] = struct{}{}
 
-		if key == "_id" {
+		if key == idKey {
 			idPresent = true
 		}
 
 		value := must.NotFail(d.Get(key))
 
-		// TODO Add dance tests for infinity: https://github.com/FerretDB/FerretDB/issues/1151
-		if v, ok := value.(float64); ok && math.IsInf(v, 0) {
-			return newValidationError(
-				ErrValidation, fmt.Errorf("invalid value: { %q: %f } (infinity values are not allowed)", key, v),
-			)
+		switch v := value.(type) {
+		case *Document:
+			err := v.ValidateData()
+			if err != nil {
+				var vErr *ValidationError
+
+				if errors.As(err, &vErr) && vErr.code == ErrIDNotFound {
+					continue
+				}
+
+				return err
+			}
+		case *Array:
+			if key == idKey {
+				return newValidationError(ErrWrongIDType, fmt.Errorf("The '_id' value cannot be of type array"))
+			}
+
+			for i := 0; i < v.Len(); i++ {
+				item := must.NotFail(v.Get(i))
+
+				switch item := item.(type) {
+				case *Document:
+					err := item.ValidateData()
+					if err != nil {
+						var vErr *ValidationError
+
+						if errors.As(err, &vErr) && vErr.code == ErrIDNotFound {
+							continue
+						}
+
+						return err
+					}
+				case *Array:
+					return newValidationError(ErrValidation, fmt.Errorf(
+						"invalid value: { %q: %v } (nested arrays are not allowed)", key, FormatAnyValue(v),
+					))
+				}
+			}
+		case float64:
+			// TODO Add dance tests for infinity: https://github.com/FerretDB/FerretDB/issues/1151
+			if math.IsInf(v, 0) {
+				return newValidationError(
+					ErrValidation, fmt.Errorf("invalid value: { %q: %f } (infinity values are not allowed)", key, v),
+				)
+			}
+		case Regex:
+			if key == idKey {
+				return newValidationError(ErrWrongIDType, fmt.Errorf("The '_id' value cannot be of type regex"))
```

Comment:
For example, should that error text use `idKey`?

Reply:
Not sure. I think no, because this is the error message and there is no other possible options.